scripts/getTsupConfig.js

```diff
@@ -0,0 +1,41 @@
+// @ts-check
+
+/**
+ * @param {Object} opts - Options for building configurations.
+ * @param {string[]} opts.entry - The entry array.
+ * @param {boolean} opts.bundle - Whether to bundle the output.
+ * @returns {import('tsup').Options}
+ */
+export function modernConfig(opts) {
+  return {
+    entry: opts.entry,
+    format: ['cjs', 'esm'],
+    target: ['chrome84', 'firefox90', 'edge84', 'safari15', 'ios15', 'opera70'],
+    outDir: 'build/modern',
```

Comment:
the modern config doesn't seem to keep optional chaining. For example, `query-core/query.js` on the alpha branch has:

```
    this.#dispatch({
      data,
      type: 'success',
      dataUpdatedAt: options?.updatedAt,
      manual: options?.manual
    });
```

while on this PR, it has:

```
    this.#dispatch({
      data,
      type: "success",
      dataUpdatedAt: options == null ? void 0 : options.updatedAt,
      manual: options == null ? void 0 : options.manual
    });
```

Reply:
and the `query.cjs` file has a lot of helpers defined at the top that we didn't have before; this is in every `.cjs` file in the `modern` directory:

![Screenshot 2023-07-15 at 09 55 02](https://github.com/TanStack/query/assets/1021430/8f6a53fa-aa87-4c9d-aa52-0e634b158370)

do you know what that's about?

Reply:
@TkDodo the optional chaining is strange - I tried changing the versions and found it works with `chrome91` and above. In the [esbuild source](https://github.com/evanw/esbuild/blob/5ede71a321130973765c7121659a094d00d407d6/internal/compat/js_table.go#L623C2-L623C2) it marks it as requiring >91, but I'll look into this.

The extra CJS helpers are generated by esbuild for all CJS files, which I think are to improve interop within the messy CJS/ESM ecosystem. From [the docs](https://esbuild.github.io/api/#platform):

```
When the format is set to cjs but the entry point is ESM, esbuild will add special annotations for any named exports to enable importing those named exports using ESM syntax from the resulting CommonJS file. Node's documentation has more information about node's detection of CommonJS named exports.
```

Reply:
@TkDodo I've figured out what's going on with optional chaining! To generate browser compat data, esbuild uses a combination of caniuse-lite, @mdn/browser-compat-data, and kangax/compat-table. Kangax has a comprehensive test suite accessible [here](https://kangax.github.io/compat-table/es2016plus/). For `javascript.operators.optional_chaining`, chrome only passed 4 out of 5 tests from versions 80 to 90, and passes all tests from version 91 onwards.

Reply:
This is the difference between the currently proposed targets and the minimum targets that pass 5/5 tests for optional chaining:
- [Before (chrome 84, edge 84, opera 70)](https://browsersl.ist/#q=%22browserslist%22%3A+%5B%22Chrome%3E%3D84+or+Firefox%3E%3D90+or+Edge%3E%3D84+or+Safari%3E%3D15+or+iOS%3E%3D15+or+opera%3E%3D70%22%5D)
- [After (chrome 91, edge 91, opera 77)](https://browsersl.ist/#q=%22browserslist%22%3A+%5B%22Chrome%3E%3D91+or+Firefox%3E%3D90+or+Edge%3E%3D91+or+Safari%3E%3D15+or+iOS%3E%3D15+or+opera%3E%3D77%22%5D)

Reply:
More context: Babel also generates its compatibility from kangax/compat-table, however the `bugfixes: true` option allows optional chaining to work from chrome 80 onwards ([plugin-bugfixes.json](https://github.com/babel/babel/blob/86fee43f499c76388cab495c8dcc4e821174d4e0/packages/babel-compat-data/data/plugin-bugfixes.json) vs [plugins.json](https://github.com/babel/babel/blob/86fee43f499c76388cab495c8dcc4e821174d4e0/packages/babel-compat-data/data/plugins.json)). AFAIK tsup/esbuild don't have an equivalent option to patch browser bugs.

Reply:
okay so lets just use the current browserslist with `bugfixes: false` then?

Reply:
> AFAIK tsup/esbuild don't have an equivalent option to patch browser bugs.

aah, I see. okay, then lets use the higher browserslist, but please also update the docs